opt_einsum/paths.py

```diff
@@ -272,3 +279,201 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
         path_cost += best[0][1]
 
     return path
+
+
+def ssa_to_linear(ssa_path):
+    """
+    Convert a path with static single assignment ids to a path with recycled
+    linear ids.
+    """
+    ids = np.arange(sum(map(len, ssa_path)), dtype=np.int64)
+    path = []
+    for i, ssa_ids in enumerate(ssa_path):
+        path.append(tuple(ids[ssa_id] for ssa_id in ssa_ids))
+        for ssa_id in ssa_ids:
+            ids[ssa_id:] -= 1
+    return path
+
+
+def linear_to_ssa(path):
+    """
+    Convert a path with recycled linear ids to a path with static single
+    assignment ids.
+    """
+    num_inputs = sum(map(len, path)) - len(path) + 1
+    linear_to_ssa = list(range(num_inputs))
+    new_ids = itertools.count(num_inputs)
+    ssa_path = []
+    for ids in path:
+        ssa_path.append(tuple(linear_to_ssa[id_] for id_ in ids))
+        for id_ in sorted(ids, reverse=True):
+            del linear_to_ssa[id_]
+        linear_to_ssa.append(next(new_ids))
+    return ssa_path
+
+
+def _get_candidate(output, sizes, remaining, dim_ref_counts, k1, k2):
+    either = k1 | k2
+    two = k1 & k2
+    one = either - two
+    k12 = (either & output) | (two & dim_ref_counts[3]) | (one & dim_ref_counts[2])
+    cost = helpers.compute_size_by_dict(k12, sizes)
```

Comment:
What was the reasoning for moving away from ``size(k12) - size(k1) - size(k2)``? It produces noticeable worse contractions for me in quite a few examples.

Reply:
Oh ok, I'll change back to a variant of `size12 - size1 - size2`. My motivation was that this objective led to more determinism in a large example I was looking at, and the increased sharing led to overall cheaper cost. I'll see if I can achieve the same effect with a different tie-breaker.

Reply:
Yes for most 'difficult' examples I've tried (e.g. contracting quantum circuits), doing it this way produces contractions with a bit worse to a lot worse scaling so my preference would be to keep it.

On a side note, it might be nice at some point to be able to supply a custom function here (e.g. just ``size12``). I particularly like the idea of supplying several even intentionally random costs and then finding the best scaling across several attempts. Might be an issue best left for later however!

Reply:
Echoing @jcmgray here, if this isn't too much of a hotspot `size(k12) - size(k1) - size(k2)` should prevent some obviously poor pathways.